utils.go

```diff
@@ -51,6 +54,19 @@ func safeAsyncClose(b *Broker) {
 	})
 }
 
+func asyncCloseWithAck(b *Broker, wg *sync.WaitGroup) {
+	tmp := b // local var prevents clobbering in goroutine
+	wg.Add(1)
+	go func () {
+		withRecover(func() {
+			if err := tmp.Close(); err != nil {
```

Comment:
broker.close() closes its network connection to the broker and triggers an shutdown of its associated routines.  Don't understand the invalid error issue.